### PR TITLE
feat(plugins): init container to download plugin resources

### PR DESCRIPTION
**Pull request by link108**

### CODE REVIEW

File: halyard-deploy/src/main/java/com/netflix/spinnaker/halyard/deploy/spinnaker/v1/profile/PluginProfileFactory.java [hunk 1]
```diff
@@ -30,11 +31,20 @@
 @Slf4j
 @Component
 public class PluginProfileFactory extends StringBackedProfileFactory {
-  @Override
-  protected void setProfile(
-      Profile profile,
-      DeploymentConfiguration deploymentConfiguration,
-      SpinnakerRuntimeSettings endpoints) {
+
+  public Profile getProfileByServiceName(
```
Comment (link108) [inline halyard-deploy/src/main/java/com/netflix/spinnaker/halyard/deploy/spinnaker/v1/profile/PluginProfileFactory.java]: I'm not the biggest fan of this, its kind of going around the whole `getProfile` / `setProfile` pattern thats going on with ProfileFactories

File: halyard-deploy/src/main/java/com/netflix/spinnaker/halyard/deploy/spinnaker/v1/profile/PluginProfileFactory.java [hunk 2]
```diff
@@ -53,17 +63,25 @@ protected void setProfile(
...
     metadata.put("enabled", plugin.getEnabled());
     metadata.put("name", manifest.getName());
-    metadata.put("jars", manifest.getJars());
+    metadata.put("jars", manifest.getResources().get(serviceName));
     metadata.put("manifestVersion", manifest.getManifestVersion());
     return metadata;
   }
 
+  @Override
+  protected void setProfile(
```
Comment (link108) [inline halyard-deploy/src/main/java/com/netflix/spinnaker/halyard/deploy/spinnaker/v1/profile/PluginProfileFactory.java:83]: see comment above

File: halyard-config/src/main/java/com/netflix/spinnaker/halyard/config/model/v1/node/Plugins.java [hunk 3]
```diff
@@ -22,10 +22,14 @@
 import java.util.stream.Collectors;
 import lombok.Data;
 import lombok.EqualsAndHashCode;
+import lombok.Getter;
 
 @Data
 @EqualsAndHashCode(callSuper = false)
 public class Plugins extends Node {
+  @Getter public boolean enabled;
```
Comment (ncknt) [inline halyard-config/src/main/java/com/netflix/spinnaker/halyard/config/model/v1/node/Plugins.java]: `@Data` or `@Getter` but not both. Also not `public`

File: halyard-config/src/main/java/com/netflix/spinnaker/halyard/config/model/v1/plugins/Manifest.java [hunk 4]
```diff
@@ -30,15 +30,15 @@
...
   public Map<String, Object> options;
+  public Map<String, List<String>> resources;
 
   static final String regex = "^[a-zA-Z0-9]+\\/[\\w-]+$";
   static final Pattern pattern = Pattern.compile(regex);
 
   public void validate() throws HalException {
 
-    if (Stream.of(name, manifestVersion, jars).anyMatch(Objects::isNull)) {
+    if (Stream.of(name, manifestVersion).anyMatch(Objects::isNull)) {
```
Comment (ncknt) [inline halyard-config/src/main/java/com/netflix/spinnaker/halyard/config/model/v1/plugins/Manifest.java]: Is this really better than `name == null || manifestVersion == null`?

File: halyard-deploy/src/main/java/com/netflix/spinnaker/halyard/deploy/spinnaker/v1/service/distributed/kubernetes/v2/KubernetesV2DeckService.java [hunk 5]
```diff
@@ -70,6 +75,94 @@ public boolean runsOnJvm() {
...
+          .getDeploymentConfiguration()
+          .getDeploymentEnvironment()
+          .getInitContainers()
+          .put(getServiceName(), initContainers);
+    }
+
+    return KubernetesV2Service.super.getInitContainers(details);
+  }
+
+  private Map getPluginDownloadingContainer(Plugins plugins) {
```
Comment (ncknt) [inline halyard-deploy/src/main/java/com/netflix/spinnaker/halyard/deploy/spinnaker/v1/service/distributed/kubernetes/v2/KubernetesV2DeckService.java]: Generic on the `Map`

File: halyard-deploy/src/main/java/com/netflix/spinnaker/halyard/deploy/spinnaker/v1/profile/PluginProfileFactory.java [hunk 6]
```diff
@@ -53,17 +63,25 @@ protected void setProfile(
 
     profile.appendContents(
         yamlToString(deploymentConfiguration.getName(), profile, fullyRenderedYaml));
+    return profile;
   }
 
-  private Map<String, Object> composeMetadata(Plugin plugin, Manifest manifest) {
+  public static Map<String, Object> composeMetadata(
```
Comment (ncknt) [inline halyard-deploy/src/main/java/com/netflix/spinnaker/halyard/deploy/spinnaker/v1/profile/PluginProfileFactory.java]: don't make it static

File: halyard-deploy/src/main/java/com/netflix/spinnaker/halyard/deploy/spinnaker/v1/service/distributed/kubernetes/v2/KubernetesV2DeckService.java [hunk 7]
```diff
@@ -70,6 +75,94 @@ public boolean runsOnJvm() {
...
+    List<String> command = new ArrayList();
+    command.add("/bin/sh");
+    return command;
+  }
+
+  private List<String> getPluginDownloadingArgs(List<Manifest> manifests) {
+    String str = "cd /opt/spinnaker/plugins/resources; ";
+    for (Manifest manifest : manifests) {
+      String manifestsList = String.join(" ", manifest.getResources().get(getServiceName()));
+      str = str.concat("mkdir -p " + manifest.getName() + "; ");
```
Comment (link108) [inline halyard-deploy/src/main/java/com/netflix/spinnaker/halyard/deploy/spinnaker/v1/service/distributed/kubernetes/v2/KubernetesV2DeckService.java]: string builder, append can do a builder pattern

**Comment by link108** [comment]
Make more generics

File: halyard-deploy/src/main/java/com/netflix/spinnaker/halyard/deploy/spinnaker/v1/service/distributed/kubernetes/v2/KubernetesV2Service.java [hunk 8]
```diff
@@ -256,7 +256,8 @@ default String getPodSpecYaml(
         .addBinding("affinity", getAffinity(details))
         .addBinding("tolerations", getTolerations(details))
         .addBinding(
-            "volumes", combineVolumes(configSources, settings.getKubernetes(), sidecarConfigs))
+            "volumes",
```
Comment (link108) [inline halyard-deploy/src/main/java/com/netflix/spinnaker/halyard/deploy/spinnaker/v1/service/distributed/kubernetes/v2/KubernetesV2Service.java]: revert these changes, we add init containers in stageConfig

File: halyard-deploy/src/main/java/com/netflix/spinnaker/halyard/deploy/spinnaker/v1/service/distributed/kubernetes/v2/KubernetesV2DeckService.java [hunk 9]
```diff
@@ -70,6 +75,100 @@ public boolean runsOnJvm() {
...
+  }
+
+  private List<String> getCommand() {
+    List<String> command = new ArrayList();
+    command.add("/bin/sh");
+    return command;
+  }
+
+  private List<String> getPluginDownloadingArgs(List<Manifest> manifests) {
+    StringBuilder sb = new StringBuilder("cd /opt/spinnaker/plugins/resources; ");
```
Comment (bpowell) [inline halyard-deploy/src/main/java/com/netflix/spinnaker/halyard/deploy/spinnaker/v1/service/distributed/kubernetes/v2/KubernetesV2DeckService.java]: instead of ending each shell command with a `;`, it might make more sense to use `&&` in case one of the commands fails, the rest of the commands will not run.

**Comment by link108** [comment]
closing to make a PR to OSS
